File: lib/gherkin/listener/formatter_listener.rb
require 'gherkin/native'
require 'gherkin/formatter/model'

module Gherkin
  module Listener
    # Adapter from the "raw" Gherkin <tt>Listener</tt> API
    # to the slightly more high-level <tt>Formatter</tt> API,
    # which is easier to implement (less state to keep track of).
    class FormatterListener
      native_impl('gherkin')

      def initialize(formatter)
        @formatter = formatter
        @stash = Stash.new
      end

      def comment(value, line)
        @stash.comment Formatter::Model::Comment.new(value, line)
      end

      def tag(name, line)
        @stash.tag Formatter::Model::Tag.new(name, line)
      end

      def feature(keyword, name, description, line)
        @stash.feature(name) do |comments, tags, id|
          replay Formatter::Model::Feature.new(comments, tags, keyword, name, description, line, id)
        end
      end

      def background(keyword, name, description, line)
        @stash.feature_element(name) do |comments, tags, id|
          replay Formatter::Model::Background.new(comments, keyword, name, description, line)
        end
      end

      def scenario(keyword, name, description, line)
        replay_step_or_examples
        @stash.feature_element(name) do |comments, tags, id|
          replay Formatter::Model::Scenario.new(comments, tags, keyword, name, description, line, id)
        end
      end

      def scenario_outline(keyword, name, description, line)
        replay_step_or_examples
        @stash.feature_element(name) do |comments, tags, id|
          replay Formatter::Model::ScenarioOutline.new(comments, tags, keyword, name, description, line, id)
        end
      end

      def examples(keyword, name, description, line)
        replay_step_or_examples
        @stash.examples(name) do |comments, tags, id|
          @current_builder = Formatter::Model::Examples::Builder.new(comments, tags, keyword, name, description, line, id)
        end
      end

      def step(keyword, name, line)
        replay_step_or_examples
        @stash.basic_statement do |comments|
          @current_builder = Formatter::Model::Step::Builder.new(comments, keyword, name, line)
        end
      end

      def row(cells, line)
        @stash.basic_statement do |comments, id|
          @current_builder.row(comments, cells, line, id)
        end
      end

      def doc_string(content_type, value, line)
        @current_builder.doc_string(value, content_type, line)
      end

      def eof
        replay_step_or_examples
        @formatter.eof
      end

      def syntax_error(state, ev, legal_events, uri, line)
        @formatter.syntax_error(state, ev, legal_events, uri, line)
      end

    private
    
      def replay(element)
        element.replay(@formatter)
      end
      
      class Stash
        attr_reader :comments, :tags, :ids
        
        def initialize
          @comments, @tags, @ids = [], [], []
          @row_index = 0
        end
        
        def comment(comment)
          @comments << comment
        end
        
        def feature(name)
          @feature_id = id(name)
          yield @comments, @tags, @feature_id
          @comments, @tags = [], []
        end

        def feature_element(name)
          @feature_element_id = "#{@feature_id};#{id(name)}"
          yield @comments, @tags, @feature_element_id
          @comments, @tags = [], []
        end
        
        def examples(name)
          @examples_id = "#{@feature_element_id};#{id(name)}"
          @row_index = 0
          yield @comments, @tags, @examples_id
          @comments, @tags = [], []
        end
        
        def basic_statement
          @row_index += 1
          yield @comments, "#{@examples_id};#{@row_index}"
          @comments = []
        end
        
        def tag(tag)
          @tags << tag
        end

        def id(name)
          (name || '').gsub(/\s/, '-').downcase
        end
      end

      def replay_step_or_examples
        return unless @current_builder
        replay(@current_builder)
        @current_builder = nil
      end
    end
  end
end

File: java/src/main/java/gherkin/listener/FormatterListener.java
package gherkin.listener;

import gherkin.Listener;
import gherkin.formatter.Formatter;
import gherkin.formatter.model.*;

import java.util.ArrayList;
import java.util.List;

public class FormatterListener implements Listener {
    private final Formatter formatter;
    private String uri;
    private List<Comment> comments = new ArrayList<Comment>();
    private List<Tag> tags = new ArrayList<Tag>();
    private Statement step;
    private List<Row> table;
    private Statement examples;
    private PyString pyString;

    public FormatterListener(Formatter formatter) {
        this.formatter = formatter;
    }

    public void location(String uri) {
        this.uri = uri;
    }

    public void comment(String comment, int line) {
        comments.add(new Comment(comment, line));
    }

    public void tag(String tag, int line) {
        tags.add(new Tag(tag, line));
    }

    public void feature(String keyword, String name, String description, int line) {
        formatter.feature(statement(grabComments(), grabTags(), keyword, name, description, line), uri);
    }

    public void background(String keyword, String name, String description, int line) {
        formatter.background(statement(grabComments(), grabTags(), keyword, name, description, line));
    }

    public void scenario(String keyword, String name, String description, int line) {
        replayStepsOrExamples();
        formatter.scenario(statement(grabComments(), grabTags(), keyword, name, description, line));
    }

    public void scenarioOutline(String keyword, String name, String description, int line) {
        replayStepsOrExamples();
        formatter.scenarioOutline(statement(grabComments(), grabTags(), keyword, name, description, line));
    }

    public void examples(String keyword, String name, String description, int line) {
        replayStepsOrExamples();
        examples = statement(grabComments(), grabTags(), keyword, name, description, line);
    }

    public void step(String keyword, String name, int line) {
        replayStepsOrExamples();
        step = statement(grabComments(), grabTags(), keyword, name, null, line);
    }

    public void row(List<String> cells, int line) {
        if (table == null) {
            table = new ArrayList<Row>();
        }
        table.add(new Row(grabComments(), cells, line));
    }

    public void pyString(String string, int line) {
        this.pyString = new PyString(string, line);
    }

    public void eof() {
        replayStepsOrExamples();
        formatter.eof();
    }

    public void syntaxError(String state, String event, List<String> legalEvents, String uri, int line) {
        formatter.syntaxError(state, event, legalEvents, uri, line);
    }

    private Statement statement(List<Comment> comments, List<Tag> tags, String keyword, String name, String description, int line) {
        return new Statement(comments, tags, keyword, name, description, line);
    }

    private List<Comment> grabComments() {
        List<Comment> comments = this.comments;
        this.comments = new ArrayList<Comment>();
        return comments;
    }

    private List<Tag> grabTags() {
        List<Tag> tags = this.tags;
        this.tags = new ArrayList<Tag>();
        return tags;
    }

    private List<Row> grabTable() {
        List<Row> table = this.table;
        this.table = null;
        return table;
    }

    private PyString grabPyString() {
        PyString pyString = this.pyString;
        this.pyString = null;
        return pyString;
    }

    private void replayStepsOrExamples() {
        if (step != null) {
            PyString pyString;
            List<Row> table;
            if ((pyString = grabPyString()) != null) {
                step.replayStep(formatter, pyString, null);
            } else if ((table = grabTable()) != null) {
                step.replayStep(formatter, table, null);
            } else {
                step.replayStep(formatter, (PyString) null, null);
            }
            step = null;
        }
        if (examples != null) {
            examples.replayExamples(formatter, grabTable());
            examples = null;
        }
    }
}
